tools/block-generator/generator/generate.go

```diff
@@ -369,90 +299,59 @@ func (g *generator) WriteBlock(output io.Writer, round uint64) error {
-			RewardsState: bookkeeping.RewardsState{
-				FeeSink:                   g.feeSink,
-				RewardsPool:               g.rewardsPool,
-				RewardsLevel:              0,
-				RewardsRate:               0,
-				RewardsResidue:            0,
-				RewardsRecalculationRound: 0,
-			},
-			UpgradeState: bookkeeping.UpgradeState{
-				CurrentProtocol: g.protocol,
-			},
-			UpgradeVote:        bookkeeping.UpgradeVote{},
-			StateProofTracking: nil,
-		}
+		g.setBlockHeader(&cert)
 
-		// Generate the transactions
-		transactions := []txn.SignedTxnInBlock{}
-		for intra < numTxnForBlock {
-			var signedTxns []txn.SignedTxn
-			var ads []txn.ApplyData
+		txGroupsAD := [][]txn.SignedTxnWithAD{}
+		for intra < minTxnsForBlock {
+			var txGroupAD []txn.SignedTxnWithAD
 			var err error
-			signedTxns, ads, intra, err = g.generateSignedTxns(g.round, intra)
+			txGroupAD, intra, err = g.generateTxGroup(g.round, intra)
 			if err != nil {
-				// return err
 				return fmt.Errorf("failed to generate transaction: %w", err)
 			}
-			if len(signedTxns) == 0 {
+			if len(txGroupAD) == 0 {
 				return fmt.Errorf("failed to generate transaction: no transactions given")
 			}
-			if len(signedTxns) != len(ads) {
-				return fmt.Errorf("failed to generate transaction: mismatched number of signed transactions (%d) and apply data (%d)", len(signedTxns), len(ads))
-			}
-			for i, stx := range signedTxns {
-				stib, err := cert.Block.BlockHeader.EncodeSignedTxn(stx, ads[i])
-				if err != nil {
-					return fmt.Errorf("failed to encode transaction: %w", err)
-				}
-				transactions = append(transactions, stib)
-			}
+			txGroupsAD = append(txGroupsAD, txGroupAD)
 		}
 
-		if intra < numTxnForBlock {
-			return fmt.Errorf("not enough transactions generated: %d > %d", numTxnForBlock, intra)
+		vBlock, ledgerTxnCount, err := g.evaluateBlock(cert.Block.BlockHeader, txGroupsAD, int(intra))
+		if err != nil {
+			return fmt.Errorf("failed to evaluate block: %w", err)
+		}
+		if ledgerTxnCount != g.txnCounter + intra {
+			return fmt.Errorf("evaluateBlock() txn count mismatches theoretical intra: %d != %d", ledgerTxnCount, g.txnCounter + intra)
 		}
 
-		cert.Block.BlockHeader.TxnCounter = g.txnCounter + intra
-		cert.Block.Payset = transactions
-		cert.Certificate = agreement.Certificate{} // empty certificate for clarity
-
-		var errs []error
-		err := g.ledger.AddBlock(cert.Block, cert.Certificate)
+		err = g.ledger.AddValidatedBlock(*vBlock, cert.Certificate)
 		if err != nil {
-			errs = append(errs, fmt.Errorf("error in AddBlock: %w", err))
+			return fmt.Errorf("failed to add validated block: %w", err)
 		}
+
+		cert.Block.Payset = vBlock.Block().Payset
+
 		if g.verbose {
-			errs2 := g.introspectLedgerVsGenerator(g.round, intra)
-			if errs2 != nil {
-				errs = append(errs, errs2...)
+			errs := g.introspectLedgerVsGenerator(g.round, intra)
+			if len(errs) > 0 {
+				return fmt.Errorf("introspectLedgerVsGenerator: %w", errors.Join(errs...))
```

Comment:
using a new go 1.20 feature: [errors.Join](https://go.dev/doc/go1.20#errors)

Reply:
Nice tip. I was concatenating error strings together this week, need to start using this.